### ISSUE TITLE
 Grille de hauteurs pour calculer la CT

### ISSUE THREAD

**Issue by benRenard** [BaRatin-tools/BaRatinAGE#19]
Il arrive que les utilisateurs ne réalisent pas que la courbe de tarage est calculée sur des pas de hauteur trop lâches. Une solution pourrait être de proposer une grille adaptative (à pas log? à pas ajustés sur la variation relative de Q Maxpost? à pas incluant les hauteurs des jaugeages?)

**Comment by JeromeLeCoz** [comment]
1. I would prefer a truely log grid, with rounded values (eg 2 or 3 significant digits). However we'll have the problem that stage can take zero and negative values... which might be managed with a combined lin/log scale indeed. Not very simple...
2. it may be easier to subsample the results of a single BaM run performed on a dense grid? The user would specify the default stage interval and the max tolerated discharge variation (in %). There could be a warning issued if the max tolerated discharge variation is exceeded between two points of the default grid.I don't see the value of including the gauging stage values in the grid (?). And it may result in irregular (ugly?) stage grids.

Yes, importing/exporting/editing the grid would be nice, I think.

**Comment by JeromeLeCoz** [comment]
1. Agreed. What about letting the user define constant grid steps over several grid ranges? eg 1 mm from hmin to h1, 1 cm from h1 to h2, 10 cm from h2 to hmax? By default, no additional breakpoint hi defined, so that we fall back on the current option.
2. Agreed. However, in many cases we wont't wish to keep all the grid points, especially when exporting (Barème allows for 100 points max, Jacinthe has no limitation). If the grid step is fine enough for low flows, it would certainly mean hundreds of useless grid points for medium and high flows... So, what about adding a simple subsampling option, base on a max maxpost Q deviation from linear interpolation? That would be easy to set up. We could also have a warning issued if that deviation is greater than x%.
3. Agreed.
4. Agreed, however subsampling the results grid based on a simple discharge deviation criterion would be helpful, I think.
5. Not sure I understand all the consequences, but I understand this would not be for near future anyway?

**Comment by IvanHeriver** [comment]
Ok, so the main ideas we have to choose from are :
1. mutli segmentent stage grid suggested by Jérôme
2. grid as a function of discharge variation (both option below are not that straigthforward to implement)
  a. requiring a pre run of the maxpost; computationnaly intensive
  b. offer a subsampling option based on the discharge variation when exporting
    - would require that the grid is precise enough to be relevant (we can change the default number of steps to be 1000 instead of 100)
    - would make export to Barème possible regardless of the number of points
3. allow importing a custom grid making it possible to do whatever you want including option 1

In my opinion, the options 2b and 3 are best.

For option 2b, I'm not sure how to interface it though. Two options I can think of:
1. add a button to create a subsampled RC component from an existing RC component which would have a custom grid (as if imported from a spreadsheet); BaM would then be re-run with this subsampled grid.
2. only affect exported values when exporting them in the table panel (definilty easier option to implement)

**Comment by JeromeLeCoz** [comment]
Perfect solution.

**Comment by JeromeLeCoz** [comment]
Not computing the formal expression of the second derivative, but computing the discrete term over a discrete (and fine) grid of h, which is easy.

Then computing the n points between hmin and hmax: I guess it would be linear spacing over a transform of [hmin;hmax], eg. using variable z=h*|d²Q/dh²|? Then computing back the h_i = z_i / |d²Q/dh²|.

We then need a smart rounding of the h_i, to avoid meaningless/impracticable precision. For instance, rounding to 10% of the h interval? plus a minimum precision (cm)?

**Comment by IvanHeriver** [comment]
Any progress on this issue @benRenard ? (By the way, a R prototype would much better than an excel one)

**Comment by benRenard** [comment]
No sorry, and I'm afraid it's unlikely to reach the top of my 2DO list anytime soon, by lack of time (and, let's be honest, of interest). @JeromeLeCoz, do you want to give a try to this smart-grid prototype ? Otherwise we should probably consider implementing only options 1 (current linear grid) and 3 (import grid) to start with.

**Comment by JeromeLeCoz** [comment]
Yes, I won't be able to provide a prototype soon either, so let's start with options 1 and 3 first, unless you Ivan are able to build a prototype for option 2?
